### ISSUE TITLE
3.12 Release

### ISSUE THREAD

**Issue by rprouse** [nunit/nunit#3221]
Issue for planning the 3.12 release. We are waiting for @jnm2's PR, then I think we can move forward with the release. I don't see any other priorities. I've gone through and cleaned up the issues tagging everything and putting them in the correct milestones. I am going to do another pass at open PR's.

**Comment by jnm2** [comment]
The failure ("Test cancelled by user") is something I've seen intermittently in our master branch for as long as I can remember. I'm struggling to figure out how my PR is interacting with it, but the good news is that it'll force us to fix it. The bad news is that I don't know how long it's going to take to unravel this. You might have to release without my PR.

What's clear according to my testing is that [this line](https://github.com/nunit/nunit/blob/9f2d8ea7c35518ee28f1255bc21aeca6d84057b3/src/NUnitFramework/tests/Constraints/DelayedConstraintTests.cs#L261) is throwing ThreadAbortException, handled [here](https://github.com/nunit/nunit/blob/9f2d8ea7c35518ee28f1255bc21aeca6d84057b3/src/NUnitFramework/framework/Internal/Results/TestResult.cs#L593) (verified by changing the message and seeing it change in the repro).

What's not clear is why a thread abort is happening at all.

**Comment by jnm2** [comment]
It appears that using `Thread.Sleep` instead of `AutoResetEventHandle.WaitOne` has fixed the problem. I don't like that I don't understand why, but it'll unblock us at least!

**Comment by jnm2** [comment]
When I run test45 using Mono 5.20 on Windows just like we do on Linux (`&"C:\Program Files\Mono\bin\mono.exe" "C:/Users/me/Source/Repos/nunit/tools/NUnit.ConsoleRunner.3.9.0/tools/nunit3-console.exe" "C:/Users/me/Source/Repos/nunit/bin/Release/net45/nunit.framework.tests.dll" "--result=C:/Users/me/Source/Repos/nunit/test-results/net45/nunit.framework.tests.xml" --process=InProcess`), the test run deadlocks.

**Comment by jnm2** [comment]
It deadlocks at different places, too! Places I keep seeing in CI...

**Comment by jnm2** [comment]
I thought running Mono directly on Windows would shorten the loop and speed me up, but our master branch deadlocks on Mono 5.20 on Windows. I opened a separate issue to come back to: https://github.com/nunit/nunit/issues/3222

**Comment by jnm2** [comment]
@rprouse I'm not going to be able to finish this before tomorrow night. Can we aim to add it in a smaller release in a month or so? In general, that would take a lot of the pressure off do that I'm not holding everyone up over my PR or having to wait six months to get the PR released.

**Comment by jnm2** [comment]
No progress this evening. (Besides confirming that skipping the DelayedConstraintTests on Mono works around it.) I'll try again tomorrow evening.

**Comment by rprouse** [comment]
@jnm2 I've been doing less than I thought I'd be able to after the surgery, so no worries. Do you think you might get to this soon, or should I look at skipping this and doing the release?

**Comment by jnm2** [comment]
Since there are folks waiting, I'd be happiest if we skipped my PR for 3.12. Would you be open to shipping again only a month or two later?

**Comment by rprouse** [comment]
There is enough work in this one PR to justify another release soon, so I would be open to that. I have a developer meetup this week and am teaching a girls learning code course, so I will need to fit in the release as I can. If you get to the PR before then, all the better. I'll do my best to get it out this week though. I've already done most of the ground-work and cleanup, I just need to pull it together.

**Comment by jnm2** [comment]
https://github.com/nunit/nunit/issues/3209 seems like something we'd want to look at in the same timeframe then, too.

**Comment by jnm2** [comment]
@rprouse #3095 is ready to merge! Also, I'd like to merge https://github.com/nunit/nunit/pull/3236 ASAP and I threw its issue in 3.12.

**Comment by rprouse** [comment]
@jnm2 I'm reviewing the two outstanding PRs then I was planning on starting the release. Is there anything else I should be waiting for?

**Comment by rprouse** [comment]
Comments on #3236. Once it is merged, I will start the release. I will have time tonight or most evenings this week.

**Comment by rprouse** [comment]
The 3.12 release is out, I just need to update the website and announce.
